### ISSUE TITLE
npcs not loading after restart in spawn chunks

### ISSUE THREAD

**Issue by BloodEko** [CitizensDev/Citizens2#1332]
Hey!
I guess there's a bug in citizens. When anyone joins too fast directly after serverRestart, npcs in his chunks aren't loaded and won't load themself later.

Maybe the player join / chunk loading before all npcs are loaded causes some internal problems ?

So I had i.e. situations where the players wondered, why relevant npcs are missing.

spigot: 1.12  citizens: latest

**Comment by mcmonkey4eva** [comment]
I've seen this one myself, actually.

**Comment by DarkryCZ** [comment]
This problem is not only after restart but after every login next to the NPC it happens too. This is the problem on the side of player, I think, because I can't see them, but another player can. after /citizens reload or going out of chunk they should be visible again, but as I say- it is only for 1 player... That means that you can reload them and then someone logs in and they are invisible again, but only for him.

**Comment by BloodEko** [comment]
@DarkryCZ 
For me, when npc's are loaded, i can relog next to them without problems. Maybe you got another issue ?

**Comment by MusicScore** [comment]
Can confirm this happens on the latest Citizens build, with the latest Spigot 1.12.2 build.

**Comment by DarkryCZ** [comment]
@BloodEko 
I checked my settings of NPCs in folders and I didn't find anything unusual. All settings were ok and their saving of positions should be ok too. Maybe it is caused by my hosting ;)

**Comment by firestrife23** [comment]
Can confirm the issue above, and here's snippet of Console Log when attempting to run command /citizens save... could be related?

`20.12 10:18:03 [Server] WARN java.lang.NullPointerException
20.12 10:18:03 [Server] WARN at net.dandielo.core.items.serialize.core.LeatherColor.serialize(LeatherColor.java:41)
20.12 10:18:03 [Server] WARN at net.dandielo.core.items.serialize.ItemAttribute.toString(ItemAttribute.java:109)
20.12 10:18:03 [Server] WARN at net.dandielo.core.items.dItem.serialize(dItem.java:115)
20.12 10:18:03 [Server] WARN at net.dandielo.citizens.traders_v3.traders.stock.StockItem.toString(StockItem.java:87)
20.12 10:18:03 [Server] WARN at net.dandielo.citizens.traders_v3.traders.stock.StockTrader.save(StockTrader.java:121)
20.12 10:18:03 [Server] WARN at net.dandielo.citizens.traders_v3.traits.TraderTrait.save(TraderTrait.java:70)
20.12 10:18:03 [Server] WARN at net.citizensnpcs.api.npc.AbstractNPC.save(AbstractNPC.java:341)
20.12 10:18:03 [Server] WARN at net.citizensnpcs.npc.CitizensNPC.save(CitizensNPC.java:150)
20.12 10:18:03 [Server] WARN at net.citizensnpcs.api.npc.SimpleNPCDataStore.store(SimpleNPCDataStore.java:80)
20.12 10:18:03 [Server] WARN at net.citizensnpcs.Citizens.storeNPCs(Citizens.java:449)
20.12 10:18:03 [Server] WARN at net.citizensnpcs.Citizens.storeNPCs(Citizens.java:454)
20.12 10:18:03 [Server] WARN at net.citizensnpcs.commands.AdminCommands.save(AdminCommands.java:64)
20.12 10:18:03 [Server] WARN at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
20.12 10:18:03 [Server] WARN at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
20.12 10:18:03 [Server] WARN at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
20.12 10:18:03 [Server] WARN at java.lang.reflect.Method.invoke(Method.java:498)
20.12 10:18:03 [Server] WARN at net.citizensnpcs.api.command.CommandManager.executeMethod(CommandManager.java:152)
20.12 10:18:03 [Server] WARN at net.citizensnpcs.api.command.CommandManager.execute(CommandManager.java:87)
20.12 10:18:03 [Server] WARN at net.citizensnpcs.api.command.CommandManager.executeSafe(CommandManager.java:175)
20.12 10:18:03 [Server] WARN at net.citizensnpcs.Citizens.onCommand(Citizens.java:237)
20.12 10:18:03 [Server] WARN at org.bukkit.command.PluginCommand.execute(PluginCommand.java:44)
20.12 10:18:03 [Server] WARN at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:151)
20.12 10:18:03 [Server] WARN at org.bukkit.craftbukkit.v1_12_R1.CraftServer.dispatchCommand(CraftServer.java:674)
20.12 10:18:03 [Server] WARN at net.minecraft.server.v1_12_R1.PlayerConnection.handleCommand(PlayerConnection.java:1471)
20.12 10:18:03 [Server] WARN at net.minecraft.server.v1_12_R1.PlayerConnection.a(PlayerConnection.java:1276)
20.12 10:18:03 [Server] WARN at net.minecraft.server.v1_12_R1.PacketPlayInChat.a(PacketPlayInChat.java:45)
20.12 10:18:03 [Server] WARN at net.minecraft.server.v1_12_R1.PacketPlayInChat.a(PacketPlayInChat.java:5)
20.12 10:18:03 [Server] WARN at net.minecraft.server.v1_12_R1.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:14)
20.12 10:18:03 [Server] WARN at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
20.12 10:18:03 [Server] WARN at java.util.concurrent.FutureTask.run(FutureTask.java:266)
20.12 10:18:03 [Server] WARN at net.minecraft.server.v1_12_R1.SystemUtils.a(SourceFile:46)
20.12 10:18:03 [Server] WARN at net.minecraft.server.v1_12_R1.MinecraftServer.D(MinecraftServer.java:843)
20.12 10:18:03 [Server] WARN at net.minecraft.server.v1_12_R1.DedicatedServer.D(DedicatedServer.java:426)
20.12 10:18:03 [Server] WARN at net.minecraft.server.v1_12_R1.MinecraftServer.C(MinecraftServer.java:767)
20.12 10:18:03 [Server] WARN at net.minecraft.server.v1_12_R1.MinecraftServer.run(MinecraftServer.java:665)
20.12 10:18:03 [Server] WARN at java.lang.Thread.run(Thread.java:745)`

**Comment by fullwall** [comment]
Ran into this one today, but couldn't reproduce it after restarting the server. Seemed to be an issue with the chunk being corrupted or something.

**Comment by Tiippex** [comment]
I still have the issue :o

**Comment by mcmonkey4eva** [comment]
@fullwall it might be from a player joining during startup sequence or similar, causing the chunks to be loaded too early (before Citizens runs the 1-tick delayed loadup task)?

**Comment by mcmonkey4eva** [comment]
#1170 is likely this same issue

**Comment by CeIebrimbor** [comment]
Confirmed on Paper 1.13.1 and Citizens2 build from Oct 14.

**Comment by CeIebrimbor** [comment]
This is **not** just from chunks being loaded at spawn before Citizens can load up, nor is it a one-off occurrence that can be fixed with the reload. 

The NPCs will intermittently disappear and reappear in a random fashion. You can use /citizens reload to make the NPCs come back, but then leaving and coming back later can and will cause the random disappearance again. I probably run the reload command 30 times a day to make these random NPCs come back.

**Comment by mcmonkey4eva** [comment]
@CeIebrimbor then your issue is _not_ the issue here and should be reported and handled separately.

**Comment by CeIebrimbor** [comment]
@mcmonkey4eva 
You closed #1170 in favor of this thread. Are you going to re-open that one? Why would you link the issues and then suggest they are not related? 🤔

**Comment by mcmonkey4eva** [comment]
@CeIebrimbor _that issue_ is equivalent to this one. The issue you just described is not.

**Comment by mcmonkey4eva** [comment]
@CeIebrimbor  There is this issue here, which describes NPCs failing to load during the startup sequence specifically for chunks that are loaded during the startup.
This is a known issue and the only thing track down is where in the code it sources from.

NPCs that disappear at any given time intermittently is a separate issue, one that has been seen many times before and each time has been a conflict with an unrelated plugin.

**Comment by CeIebrimbor** [comment]
I guess I will yield that my "experience" is consistent with the OP of this report made a year ago. I wouldn't hold my breath by having a separate report on the same topic after this one has been open this long.

I will post my report if the behavior continues after this year-old bug is addressed. Since my NPCs are also at spawn. 🤷‍♂️

**Comment by SXRWahrheit** [comment]
@CeIebrimbor if you have issues with non-spawn chunks and despawning NPCs, you should create a new issue to report the problem so it can be investigated.

**Comment by CeIebrimbor** [comment]
> Since my NPCs are also at spawn.

😛

**Comment by MusicScore** [comment]
But the NPCs also disappear in non-spawn chunks, do they not? If so, make another issue.

**Comment by mcmonkey4eva** [comment]
The difference is your NPCs are disappearing outside of the startup cycle.

This issue is regarding NPCs not loading during the startup cycle.

More on topic:
I still have a (mostly inactive) server that can replicate the the main issue here. I've confirmed it's still present on latest Citizens.

**Comment by SXRWahrheit** [comment]
Yes, I read your comment. But you could, ostensibly, test whether the issue you have occurs with non-spawn NPCs, if you're so sure that it's a different issue. After all, you said:

`I think there is a global, non-specific issue with NPCs despawning`

Please don't mock people who are trying to help you.

**Comment by CeIebrimbor** [comment]
My NPCs are exclusively at spawn. Just to set that straight. 

The use of the word "global" was suggesting that the issue of vanishing NPCs is likely not tied specifically to startup, but instead was wider spread given the fact that there have been 4+ similar reports.

Not mocking, just having fun because I just said they were at spawn in the post right before yours. 😜

**Comment by fullwall** [comment]
I note that you are using Paper @CeIebrimbor. Is async chunk loading enabled? If you disable async chunk loading, is the problem fixed?

**Comment by CeIebrimbor** [comment]
I suppose I can test at some point. Spigot was async before recent versions and will return to async, so not sure the problem should be associated with Paper specifically.

Minecraft is moving towards async chunk gen. This is reality.

I’ll get back with findings later on.

**Comment by Phyther** [comment]
Out of curiosity, I am on paper 1.13.2 build 408 and Citizens 2.0.24-SNAPSHOT build 1590 and am having this issue. Was a resolution ever found?

**Comment by Phyther** [comment]
I am on paper 1.13.2 build 417, and have tried with both async loading on and off. and changed my VBO setting on and off. Some NPCs appear at spawn sometimes, while others do not. Every restart I have to do /citizens reload for all of them to appear. There are no errors in the starup log.
**Plugins**
```
AdvancedAchievements, AntiMinechat*, BlocksHub, BossShopPro, BungeeTabListPlus*, Citizens, CitizensCMD, CMI, CMIEInjector*, CoreProtect, CrazyCrates, DeluxeMenus*, EnjinMinecraftPlugin, FastAsyncVoxelSniper, FastAsyncWorldEdit*, GuiShopManager*, HeadDatabase*, Jobs, JukeBox, LockWarp*, LuckPerms*, LuckPerms-GUI*, mcMMO, NoCreativePhantoms, NoteBlockAPI, PlaceholderAPI, ProtocolLib*, PyrCore, PyrParticles, QuestCreator, QuickShop, QuickSwitch*, Towny*, TownyGUI*, TownyNameUpdater*, TradeMe, Vault*, VentureChat*, Votifier, WorldEdit, WorldGuard
```
**Debug File Info**
```
2018-11-05 11:16:17 [INFO] Unable to find Rhino classes - javascript scripts won't see non-CraftBukkit classes
```

**Comment by Phyther** [comment]
Don't know if this is related, but I just got this error message in console:
```
[11:18:17] [Server thread/INFO]: Phyther issued server command: /citizens reload
[11:18:17] [Server thread/WARN]: java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Boolean
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.Settings$Setting.asBoolean(Settings.java:140)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.npc.ai.CitizensNavigator.<init>(CitizensNavigator.java:39)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.npc.CitizensNPC.<init>(CitizensNPC.java:51)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.npc.CitizensNPCRegistry.getByType(CitizensNPCRegistry.java:103)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.npc.CitizensNPCRegistry.createNPC(CitizensNPCRegistry.java:50)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.api.npc.SimpleNPCDataStore.loadInto(SimpleNPCDataStore.java:56)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.Citizens.reload(Citizens.java:362)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.commands.AdminCommands.reload(AdminCommands.java:44)
[11:18:17] [Server thread/WARN]: 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[11:18:17] [Server thread/WARN]: 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[11:18:17] [Server thread/WARN]: 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[11:18:17] [Server thread/WARN]: 	at java.lang.reflect.Method.invoke(Method.java:498)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.api.command.CommandManager.executeMethod(CommandManager.java:155)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.api.command.CommandManager.execute(CommandManager.java:87)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.api.command.CommandManager.executeSafe(CommandManager.java:183)
[11:18:17] [Server thread/WARN]: 	at net.citizensnpcs.Citizens.onCommand(Citizens.java:238)
[11:18:17] [Server thread/WARN]: 	at org.bukkit.command.PluginCommand.execute(PluginCommand.java:44)
[11:18:17] [Server thread/WARN]: 	at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:149)
[11:18:17] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_13_R2.CraftServer.dispatchCommand(CraftServer.java:729)
[11:18:17] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PlayerConnection.handleCommand(PlayerConnection.java:1761)
[11:18:17] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PlayerConnection.a(PlayerConnection.java:1565)
[11:18:17] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PacketPlayInChat.a(PacketPlayInChat.java:45)
[11:18:17] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PacketPlayInChat.a(PacketPlayInChat.java:5)
[11:18:17] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:12)
[11:18:17] [Server thread/WARN]: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[11:18:17] [Server thread/WARN]: 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[11:18:17] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.SystemUtils.a(SystemUtils.java:109)
[11:18:17] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.MinecraftServer.b(MinecraftServer.java:1016)
[11:18:17] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.DedicatedServer.b(DedicatedServer.java:439)
[11:18:17] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:943)
[11:18:17] [Server thread/WARN]: 	at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:841)
[11:18:17] [Server thread/WARN]: 	at java.lang.Thread.run(Thread.java:748)
[11:18:33] [Async Chat Thread - #0/INFO]: [Wanderer] Heliopolis danphanto ?  sounds good, I don't mind if you need to restart while I'm here
```

**Comment by CeIebrimbor** [comment]
I just set my tasks to run /citizens2 reload every 5 minutes because players were constantly asking "Where did the Shopkeeper go?".

**Comment by arboriginal** [comment]
Hi there,

I've juste added citizens 2 to my plugins a few hours ago and seen the same as described here. After a few tests, it seems this happens when a player log in just after the server restart, if he log in after one minute, it's ok. So, I suppose if a player connects early, the plugin process has not finished getting skins and make the NPC invisibles.

Hope this will help to find a way to fix it.

**Comment by Ghost-chu** [comment]
> 使用论文1.13.2在这里遇到同样的问题。每次服务器重新启动时，我的大多数NPC都会消失。我启用了调试以查看是否可以获得一个想法。结果：
> 
> 28.11 22:46:04 [服务器]信息由于卸载了块，重试3的生成。Util.isLoaded true 
> 28.11 22:46:04 [服务器] INFO存储3来自NPCNeedsRespawnEvent的respawn
> 
> 跑步/市民重新加载NPC回来：
> 
> 28.11 22:54:37 [服务器] INFO在已经绝望的情况下试图消灭3。
> 28.11 22:54:37 [服务器] INFO防止由于DespawnReason.RELOAD 
> 28.11 22:54:37 进一步重生2 [服务器] INFO Despawned 2 DespawnReason。RELOAD 
> 28.11 22:54:37 [服务器] INFO试图在已经绝望的情况下消灭1。
> 28.11 22:54:37 [服务器] INFO在已经堕落的情况下试图消失0。
> 
> 所以我把一个问题缩小到Paper的“keep-spawn-loaded”设置。一旦我禁用了这个，只要你在spawn中重启/重新加入，spawn中的NPC就能够重新生成每次重启。现在，如果你离开spawn，大块卸载并且NPC随之卸载并且除非你重新加载公民，否则不会回来。
> 
> 现在，我启用了Citizens：chunk - 始终保持加载，无论位置和重新启动，它都完全修复了生成的npc问题。这些都没有意义，但这是我做的测试和我从中得到的结果。我希望能找到解决方案

Failed in my test

**Comment by mcmonkey4eva** [comment]
I found a bug that seems to induce the issue, and fixed it in PR #1625 
I confirmed consistent replication of the bug on unpatched version, and consistent valid spawning on the patched version (in the PR).
When the PR is pulled (pending review) we can confirm whether the bug is patched for all other users here as well.

**Comment by mcmonkey4eva** [comment]
Fix is in build DEV-1593+
Please let me know (reply below) if any issues related to this remain on the new build!

**Comment by Agaloth** [comment]
Still happening for me (if this counts as a necropost sorry)

**Comment by mcmonkey4eva** [comment]
You are not encountering an issue that was tracked down and resolved in 2018. If you have a similar issue, it is a new one, so post a new issue for it.
